Add -O3 to jit compiler options for cpu

tcmoore3 opened the pull request:
## Description

Adds an optimization flag to the cpu compile options for the user-defined HPMC pair potentials.

## Motivation and context
No optimizations are currently being passed to the JIT compiler for pair potentials, causing poor performance when using the `hpmc.pair.user` classes.


## How has this been tested?
For 1672 hexagons with(out) square-well patches at 80% packing fraction, with mpi and 16 ranks, with docker image  `bridges2` unless otherwise noted, and not including compilation time:
| hoomd version | patches? | mean tps|
| -- | -- | -- | 
| 2.9.6 | no | 2927|
|2.9.6 | yes | 1233|
|3.1.0|no|2300|
|3.1.0|yes|143 (191 on `cheme-hodges`)|
|3.1.0 with "-O3" for patch code compilation | yes | 1169 (on `cheme-hodges`)|
|3.1.0 with "-O3 -funroll-loops" for patch code compilation | yes | 1166 (on `cheme-hodges`)|
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
- Add optimization flag to compiler options for user-defined pair potentials in HPMC
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

tcmoore3 committed:
Add -O3 to jit compiler options for cpu

tommy-waltmann reviewed (approved):
Thanks for this! It's really crazy how just adding a flag can just make the performance 10x faster

joaander commented on hoomd/_compile.py:
```diff
@@ -23,7 +23,7 @@ def _get_hoomd_include_path():
 
 def get_cpu_include_options():
     """Get the source code include path for HOOMD's include files."""
-    return ['-I', str(_get_hoomd_include_path())]
```
`-O3` is an optimization option, not an include option. Rename the method to `get_cpu_compiler_arguments` and update the docstring accordingly.

joaander reviewed (changes requested):
My apologies. All the information I read indicated that this would not work, so I didn't try it. The documentation and answers on stackoverflow say that optimization is done at the LLVM level, not clang and optimization passes need to be added on the module:

https://releases.llvm.org/13.0.1/docs/tutorial/MyFirstLanguageFrontend/LangImpl04.html#llvm-optimization-passes
https://stackoverflow.com/questions/24088403/compile-using-clang-optimisation-passes-flags

I thought I had left a TODO in the code to check the performance and add optimization passes later, but that seems to have been lost.

Does performance improve at all with `-funroll-loops`? Your code loops over a fixed number of patches, so that might help.

tcmoore3 commented:
> Does performance improve at all with `-funroll-loops`? Your code loops over a fixed number of patches, so that might help.

It does not, at least for this system. I updated the table in the description to show this.

tcmoore3 committed:
Rename get_cpu_include_options

Now that we pass the optimization flag, the new name
get_cpu_compiler_arguments is more descriptive of the function.

joaander reviewed (approved):
Thanks!

tcmoore3 committed:
Remove comments about running optimization passes

tcmoore3 committed:
Fix call to nonexistent get_cpu_include_options()
